flogger/flogger-api-tests/src/test/kotlin/com/google/common/flogger/AbstractLoggerSpec.kt

```diff
@@ -0,0 +1,163 @@
+import io.kotest.matchers.string.shouldContain
+import io.kotest.matchers.string.shouldMatch
+import io.kotest.matchers.string.shouldNotContain
+import io.spine.testing.logging.tapConsole
+import kotlin.text.RegexOption.DOT_MATCHES_ALL
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Test
+
+/**
+ * Tests for [AbstractLogger].
+ *
+ * An abstract logger is a factory for instances of fluent logging APIs,
+ * used to build log statements via method chaining.
+ *
+ * See [LogContextSpec] for the most tests related to base logging behavior.
+ *
+ * @see <a href="https://github.com/google/flogger/blob/master/api/src/test/java/com/google/common/flogger/AbstractLoggerTest.java">
+ *     Original Java code of Google Flogger</a>
+ */
+@DisplayName("`AbstractLogger` should")
+internal class AbstractLoggerSpec {
+
+    private val backend = MemoizingBackend()
+    private val logger = TestLogger.create(backend)
+
+    companion object {
+
+        /**
+         * Matches ISO 8601 date/time format.
+         *
+         * @see <a href="https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html">SimpleDateFormat</a>
+         */
+        private val ISO_TIMESTAMP_PREFIX =
+            "\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}\\.\\d{3}[-+]\\d{4}: .*".toRegex(
+                DOT_MATCHES_ALL // Makes a dot match a line terminator as well.
+            )
+    }
+
+    @Test
+    fun `report exceptions`() {
+        val bad: Any = object : Any() {
+            override fun toString(): String = error("Ooopsie")
+        }
+        val output = tapConsole {
+            logger.atInfo().log("bad value: %s", bad)
+        }
+        backend.logged.shouldBeEmpty()
+        output shouldMatch ISO_TIMESTAMP_PREFIX
+        output shouldContain "logging error"
+        output shouldContain "com.google.common.flogger.AbstractLoggerSpec.report exceptions"
+        output shouldContain "java.lang.IllegalStateException: Ooopsie"
+    }
+
+    @Test
+    fun `report nested exceptions`() {
+        // A worst case scenario whereby an object's `toString()` throws an exception,
+        // which itself throws an exception. If we can handle this, we can handle
+        // just about anything!
+        val bad: Any = object : Any() {
+            override fun toString(): String = throw object : IllegalStateException("Ooopsie") {
+                private val serialVersionUID: Long = -5383608141374997920L
+                override fun toString(): String = error("<<IGNORED>>")
+            }
+        }
+        val output = tapConsole {
+            logger.atInfo().log("bad value: %s", bad)
+        }
+        backend.logged.shouldBeEmpty()
+        output shouldMatch ISO_TIMESTAMP_PREFIX
+        output shouldContain "logging error"
+        output shouldContain "Ooopsie" // Exception class is not printed in this case.
+        output shouldNotContain "<<IGNORED>>" // Also, we don't handle the inner exception.
```

Comment:
In this test class, the code also relies on some human properly using copy-paste shortcuts — instead of having re-used `val` constants.